### PR TITLE
feat[ir]: add `make_ssa` pass to venom pipeline

### PR DESCRIPTION
**Pull request by harkal**
### Commit message

```
this commit updates the venom pipeline to be capable of translating 100%
of the original vyper IR, and successfully passes the entire test suite.

to accomplish this, the translation pass from the old IR to Venom is
simplified, moving several optimization and analysis steps to separate
passes. the most significant of these is the `make_ssa` pass, which
converts any Venom code into SSA form, therefore letting us write
non-SSA code in the translation pass, simplifying the translation.

to support the `make_ssa` pass, this commit also adds a dominator tree
implementation, along with implementations of dominance frontier and
other utility functions. these should also be useful for additional
passes that will be contributed in the future.

to facilitate the development process, this commit also adds two more
output formats: `cfg` and `cfg_runtime`, which provide a graph
representation of the Venom code.
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: tests/unit/compiler/venom/test_call.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import pytest
+
+
+@pytest.fixture
+def market_maker(get_contract):
+    contract_code = """
+from ethereum.ercs import IERC20
+
+unused: public(uint256)
```
Comment (charles-cooper) [inline tests/unit/compiler/venom/test_call.py:9]: ?

File: tests/unit/compiler/venom/test_call.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+token_address: IERC20
```
Comment (charles-cooper) [inline tests/unit/compiler/venom/test_call.py:10]: i think we can just call this `token`, because it's an IERC20 rather than an address

File: tests/unit/compiler/venom/test_dominator_tree.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+from vyper.venom.function import IRFunction
+from vyper.venom.passes.make_ssa import MakeSSA
+
+
+def _add_bb(
+    ctx: IRFunction, label: IRLabel, cfg_outs: [IRLabel], bb: Optional[IRBasicBlock] = None
+) -> IRBasicBlock:
+    bb = bb if bb is not None else IRBasicBlock(label, ctx)
+    ctx.append_basic_block(bb)
+    cgf_outs_len = len(cfg_outs)
```
Comment (charles-cooper) [inline tests/unit/compiler/venom/test_dominator_tree.py:17]: cfg

File: tests/unit/compiler/venom/test_dominator_tree.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+def test_phi_placement():
+    ctx = _make_test_ctx()
+    bb1, bb2, bb3, bb4, bb5, bb6, bb7 = [ctx.get_basic_block(str(i)) for i in range(1, 8)]
+
+    x = IRVariable("%x")
+    bb1.insert_instruction(IRInstruction("mload", [IRLiteral(0)], x), 0)
+    bb2.insert_instruction(IRInstruction("add", [x, IRLiteral(1)], x), 0)
+    bb7.insert_instruction(IRInstruction("mstore", [x, IRLiteral(0)]), 0)
+
+    MakeSSA.run_pass(ctx, bb1)
```
Comment (charles-cooper) [inline tests/unit/compiler/venom/test_dominator_tree.py:72]: do we need to test the result of run_pass?

File: tests/unit/compiler/venom/test_duplicate_operands.py [hunk 5]
```diff
@@ -23,5 +23,5 @@ def test_duplicate_operands():
     bb.append_instruction("stop")
 
     asm = generate_assembly_experimental(ctx, optimize=OptimizationLevel.GAS)
-
-    assert asm == ["PUSH1", 10, "DUP1", "DUP1", "DUP1", "ADD", "MUL", "STOP"]
+    print(asm)
```
Comment (charles-cooper) [inline tests/unit/compiler/venom/test_duplicate_operands.py:26]: stray print

File: tests/unit/compiler/venom/test_liveness_simple_loop.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    b: uint256 = 10
+
+    for i: uint256 in range(128):
+        b = 1 + _numBids
+"""
+
+
+def test_liveness_simple_loop():
+    vyper.compile_code(source, ["opcodes"])
+    assert True
```
Comment (charles-cooper) [inline tests/unit/compiler/venom/test_liveness_simple_loop.py:16]: seems unnecessary

File: tests/unit/compiler/venom/test_liveness_simple_loop.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    _numBids: uint256 = 20
+    b: uint256 = 10
+
+    for i: uint256 in range(128):
+        b = 1 + _numBids
+"""
+
+
+def test_liveness_simple_loop():
+    vyper.compile_code(source, ["opcodes"])
```
Comment (charles-cooper) [inline tests/unit/compiler/venom/test_liveness_simple_loop.py:15]: this is not running with experimental_codegen=True

File: vyper/ir/optimizer.py [hunk 8]
```diff
@@ -441,7 +441,6 @@ def _optimize(node: IRnode, parent: Optional[IRnode]) -> Tuple[bool, IRnode]:
     annotation = node.annotation
     add_gas_estimate = node.add_gas_estimate
     is_self_call = node.is_self_call
-    passthrough_metadata = node.passthrough_metadata
```
Comment (charles-cooper) [inline vyper/ir/optimizer.py:444]: oops

File: vyper/venom/venom_to_assembly.py [hunk 9]
```diff
@@ -408,34 +496,54 @@ def _generate_evm_for_instruction(
...
 
     def swap(self, assembly, stack, depth):
+        # Swaps of the top is no op
         if depth == 0:
             return
+
+        inst = _evm_swap_for(depth)
+
+        # Double swaps cancel each other out
+        if len(assembly) > 0 and inst == assembly[-1]:
```
Comment (charles-cooper) [inline vyper/venom/venom_to_assembly.py:540]: this seems like it could be an assembly peephole optimization

File: vyper/venom/venom_to_assembly.py [hunk 10]
```diff
@@ -40,35 +56,66 @@
...
 )
 
 _REVERT_POSTAMBLE = ["_sym___revert", "JUMPDEST", *PUSH(0), "DUP1", "REVERT"]
 
 
+def apply_line_numbers(inst: IRInstruction, asm) -> list[str]:
+    ret = []
+    for op in asm:
+        if isinstance(op, str) and not isinstance(op, Instruction):
+            ret.append(Instruction(op, inst.source_pos, inst.error_msg))
```
Comment (charles-cooper) [inline vyper/venom/venom_to_assembly.py:113]: this should be .ast_source now

File: vyper/venom/venom_to_assembly.py [hunk 11]
```diff
@@ -139,7 +186,11 @@ def generate_evm(self, no_optimize: bool = False) -> list[str]:
                     label = inst.operands[0].value
                     data_segments[label] = [DataHeader(f"_sym_{label}")]
                 elif inst.opcode == "db":
```
Comment (charles-cooper) [inline vyper/venom/venom_to_assembly.py:188]: why don't we have two opcodes here -- dblabel and dbdata?

File: vyper/venom/venom_to_assembly.py [hunk 12]
```diff
@@ -258,12 +331,15 @@ def clean_stack_from_cfg_in(
...
-                stack.swap(depth)
+                self.swap(asm, stack, depth)
             self.pop(asm, stack)
 
     def _generate_evm_for_instruction(
-        self, assembly: list, inst: IRInstruction, stack: StackModel
+        self, inst: IRInstruction, stack: StackModel, next_liveness: OrderedSet = None
     ) -> list[str]:
+        assembly: list[str | int] = []
+        if next_liveness is None:
```
Comment (charles-cooper) [inline vyper/venom/venom_to_assembly.py:341]: style: `next_liveness = next_liveness or OrderedSet()`

File: vyper/venom/venom_to_assembly.py [hunk 13]
```diff
@@ -277,9 +353,17 @@ def _generate_evm_for_instruction(
...
+            addr = inst.operands[0]
+            if isinstance(addr, IRLiteral):
+                operands = []
+            else:
+                operands = inst.operands
         elif opcode == "istore":
-            operands = inst.operands[0:1]
+            addr = inst.operands[1]
+            if isinstance(addr, IRLiteral):
+                operands = inst.operands[0:1]
```
Comment (charles-cooper) [inline vyper/venom/venom_to_assembly.py:364]:
```suggestion
                operands = inst.operands[:1]
```

Comment (charles-cooper) [reply vyper/venom/venom_to_assembly.py:364]: also, what is the explanation for this branch?

File: vyper/venom/venom_to_assembly.py [hunk 14]
```diff
@@ -301,7 +385,7 @@ def _generate_evm_for_instruction(
                 stack.poke(0, ret)
             else:
                 stack.poke(depth, ret)
-            return assembly
+            return apply_line_numbers(inst, assembly)
```
Comment (charles-cooper) [inline vyper/venom/venom_to_assembly.py:388]: i feel like apply_line_numbers could just be applied globally in the main venom_to_assembly routine

File: vyper/venom/passes/make_ssa.py [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+class MakeSSA(IRPass):
+    """
+    This pass converts the function into Static Single Assignment (SSA) form.
+    """
+
+    dom: DominatorTree
+    defs: dict[IRVariable, OrderedSet[IRBasicBlock]]
+
+    def _run_pass(self, ctx: IRFunction, entry: IRBasicBlock) -> int:
+        global count
```
Comment (charles-cooper) [inline vyper/venom/passes/make_ssa.py:18]: why global? i think we can use `self.count` instead?

File: vyper/venom/dominators.py [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+    """
+
+    ctx: IRFunction
+    entry: IRBasicBlock
+    dfs_order: dict[IRBasicBlock, int]
+    dfs: list[IRBasicBlock]
+    dominators: dict[IRBasicBlock, OrderedSet[IRBasicBlock]]
+    idoms: dict[IRBasicBlock, IRBasicBlock]
+    dominated: dict[IRBasicBlock, OrderedSet[IRBasicBlock]]
+    df: dict[IRBasicBlock, OrderedSet[IRBasicBlock]]
```
Comment (charles-cooper) [inline vyper/venom/dominators.py:19]:
```suggestion
    entry_block: IRBasicBlock
    dfs_index: dict[IRBasicBlock, int]
    descendants: list[IRBasicBlock]  # dfs traversal of pointed-to blocks
    dominators: dict[IRBasicBlock, OrderedSet[IRBasicBlock]]
    immediate_dominators: dict[IRBasicBlock, IRBasicBlock]
    dominated: dict[IRBasicBlock, OrderedSet[IRBasicBlock]]
    frontier: dict[IRBasicBlock, OrderedSet[IRBasicBlock]]
```

File: vyper/venom/function.py [hunk 17]
```diff
@@ -27,6 +30,10 @@ class IRFunction:
     last_label: int
     last_variable: int
 
+    # Used during code generation
+    _source_pos: list[int]
```
Comment (charles-cooper) [inline vyper/venom/function.py:34]: `ast_source: vy_ast.VyperNode`

File: vyper/venom/function.py [hunk 18]
```diff
@@ -27,6 +30,10 @@ class IRFunction:
...
+    _error_msg: list[str]
```
Comment (charles-cooper) [inline vyper/venom/function.py:35]:
```suggestion
    _error_msg: str
```

File: vyper/venom/passes/dft.py [hunk 19]
```diff
@@ -18,9 +17,9 @@ def _process_instruction_r(self, bb: IRBasicBlock, inst: IRInstruction):
             bb.instructions.append(inst)
             return
 
-        for op in inst.get_inputs():
+        for op in inst.liveness:
             target = self.dfg.get_producing_instruction(op)
-            if target.parent != inst.parent or target.fence_id != inst.fence_id:
+            if target is None or target.parent != inst.parent or target.fence_id != inst.fence_id:
```
Comment (charles-cooper) [inline vyper/venom/passes/dft.py:22]: when can target be None?

File: vyper/venom/venom_to_assembly.py [hunk 20]
```diff
@@ -139,7 +186,11 @@ def generate_evm(self, no_optimize: bool = False) -> list[str]:
                     label = inst.operands[0].value
                     data_segments[label] = [DataHeader(f"_sym_{label}")]
                 elif inst.opcode == "db":
-                    data_segments[label].append(f"_sym_{inst.operands[0].value}")
+                    data = inst.operands[0].value
+                    if isinstance(data, bytes):
+                        data_segments[label].append(data)
+                    else:
+                        data_segments[label].append(f"_sym_{inst.operands[0].value}")
```
Comment (charles-cooper) [inline vyper/venom/venom_to_assembly.py:193]: i think these cases should match the cases in `ir_node_to_venom`

File: vyper/venom/passes/make_ssa.py (hunk 15 above)
Comment (harkal) [reply vyper/venom/passes/make_ssa.py:18]: it was a debugging leftover. Not used.

File: vyper/venom/function.py (hunk 18 above)
Comment (harkal) [reply vyper/venom/function.py:35]: it's a stack of error_msgs

File: vyper/venom/__init__.py [hunk 21]
```diff
@@ -38,6 +42,24 @@ def generate_assembly_experimental(
 def _run_passes(ctx: IRFunction, optimize: OptimizationLevel) -> None:
     # Run passes on Venom IR
     # TODO: Add support for optimization levels
+
+    ir_pass_optimize_empty_blocks(ctx)
+    ir_pass_remove_unreachable_blocks(ctx)
+
+    internals = [
```
[the remark this answers is not in the document]
Comment (harkal) [reply vyper/venom/__init__.py:49]: not yet, soon!

File: vyper/venom/passes/dft.py (hunk 19 above)
Comment (harkal) [reply vyper/venom/passes/dft.py:22]: Should not happen with frontend generated code, but can happen with user generated code. Upgraded the check to assert, and in the future when we can type in venom directly we should make this a compiler error.

File: vyper/venom/passes/simplify_cfg.py [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+from vyper.venom.basicblock import IRBasicBlock
+from vyper.venom.function import IRFunction
+from vyper.venom.passes.base_pass import IRPass
+
+
+class SimplifyCFGPass(IRPass):
+    visited: OrderedSet
+
+    def _merge_blocks(self, a: IRBasicBlock, b: IRBasicBlock):
+        a.instructions.pop()
```
Comment (charles-cooper) [inline vyper/venom/passes/simplify_cfg.py:11]: i think this can be simplified right? like 
```python
a_phis = [s for s in a.instructions if s.opcode == "phi"]
b_phis = [s for s in b.instructions if s.opcode == "phi"]
a_insts = [s for s in a.instructions if s.opcode != "phi"]
b_insts = [s for s in b.instructions if s.opcode != "phi"]
b.instructions = a_phis + b_phis + a_insts + b_insts
```

File: vyper/venom/venom_to_assembly.py [hunk 23]
```diff
@@ -408,32 +503,44 @@ def _generate_evm_for_instruction(
...
             assembly.extend([f"LOG{log_topic_count}"])
         else:
             raise Exception(f"Unknown opcode: {opcode}")
 
         # Step 6: Emit instructions output operands (if any)
         if inst.output is not None:
-            assert isinstance(inst.output, IRVariable), "Return value must be a variable"
-            if inst.output.mem_type == MemType.MEMORY:
-                assembly.extend([*PUSH(inst.output.mem_addr)])
+            if "call" in inst.opcode and inst.output not in next_liveness:
```
Comment (charles-cooper) [inline vyper/venom/venom_to_assembly.py:530]: is the "call" check really necessary here?

File: vyper/venom/ir_node_to_venom.py [hunk 24]
```diff
@@ -486,334 +373,143 @@ def _convert_ir_bb(ctx, ir, symbols, variables, allocated_variables):
...
-        return ctx.get_basic_block().append_instruction(ir.value, arg_0)
-    elif ir.value in ["sstore", "istore"]:
-        arg_0, arg_1 = _convert_ir_bb_list(ctx, ir.args, symbols, variables, allocated_variables)
-        ctx.get_basic_block().append_instruction(ir.value, arg_1, arg_0)
-    elif ir.value == "unique_symbol":
-        sym = ir.args[0]
-        new_var = ctx.get_next_variable()
-        symbols[f"&{sym.value}"] = new_var
-        return new_var
+        expanded = IRnode.from_list(["with", "b", b, ["xor", "b", ["mul", cond, ["xor", a, "b"]]]])
```
Comment (charles-cooper) [inline vyper/venom/ir_node_to_venom.py:451]: order of eval -- this should really be `["with", "cond", cond, ["with", "a", a, ["with", "b" b, ["xor" ... ]]]]`

File: vyper/venom/passes/simplify_cfg.py (hunk 22 above)
Comment (harkal) [reply vyper/venom/passes/simplify_cfg.py:11]: I think just iterating over the second block once is clearer and faster than doing multiple iterations and conditions here. I vote to keep it as is.

File: vyper/venom/venom_to_assembly.py (hunk 23 above)
Comment (harkal) [reply vyper/venom/venom_to_assembly.py:530]: Yes, this is required because call can produce a value that the code might not be consuming

File: vyper/codegen/stmt.py [hunk 25]
```diff
@@ -59,6 +59,7 @@ def parse_AnnAssign(self):
         rhs = Expr(self.stmt.value, self.context).ir_node
 
         lhs = IRnode.from_list(alloced, typ=ltyp, location=MEMORY)
+        lhs.passthrough_metadata["alloca"] = (varname, alloced, ltyp.memory_bytes_required)
```
Comment (charles-cooper) [inline vyper/codegen/stmt.py:62]: stray alloca
```suggestion
```

File: vyper/ir/compile_ir.py [hunk 26]
```diff
@@ -201,9 +201,11 @@ def apply_line_no_wrapper(*args, **kwargs):
         ret = func(*args, **kwargs)
 
         new_ret = [
-            Instruction(i, code.ast_source, code.error_msg)
-            if isinstance(i, str) and not isinstance(i, Instruction)
-            else i
+            (
```
Comment (charles-cooper) [inline vyper/ir/compile_ir.py:204]: linter, this doesn't need to be changed

File: vyper/venom/venom_to_assembly.py (hunk 23 above)
Comment (charles-cooper) [reply vyper/venom/venom_to_assembly.py:530]: but why limit to `call` opcodes?

Comment (charles-cooper) [reply vyper/venom/venom_to_assembly.py:530]: should be any instruction whose output is not live.

File: setup.cfg [hunk 27]
```diff
@@ -31,3 +31,7 @@ testpaths = tests
 xfail_strict = true
 markers =
 	fuzzing: Run Hypothesis fuzz test suite (deselect with '-m "not fuzzing"')
+	venom_xfail: mark a test case as a regression (expected to fail) under the venom pipeline
+
+[tool:mypy]
+ignore_missing_imports = True
```
Comment (charles-cooper) [inline setup.cfg:37]: looks like a bad merge

File: .github/workflows/test.yml [hunk 28]
```diff
@@ -108,12 +117,14 @@ jobs:
             opt-mode: gas
             debug: false
             evm-version: shanghai
-          - python-version: ["3.12", "312"]
+
+          - python-version: ["3.11", "311"]
```
Comment (charles-cooper) [inline .github/workflows/test.yml:121]: another bad merge -- should be 3.12

File: tests/unit/compiler/venom/test_call.py [hunk 29]
```diff
@@ -0,0 +1,44 @@
+import pytest
```
Comment (charles-cooper) [inline tests/unit/compiler/venom/test_call.py:1]: is this file really needed? these should be tested in the examples tests

File: tests/unit/compiler/venom/test_multi_entry_block.py [hunk 30]
```diff
@@ -132,7 +132,7 @@ def test_multi_entry_block_with_dynamic_jump():
     assert ctx.normalized, "CFG should be normalized"
 
     finish_bb = ctx.get_basic_block(finish_label.value)
-    cfg_in = list(finish_bb.cfg_in.keys())
+    cfg_in = list(finish_bb.cfg_in)
     assert cfg_in[0].label.value == "target", "Should contain target"
-    assert cfg_in[1].label.value == "finish_split___global", "Should contain finish_split___global"
-    assert cfg_in[2].label.value == "finish_split_block_1", "Should contain finish_split_block_1"
+    assert cfg_in[1].label.value == "__global_split_finish", "Should contain __global_split_finish"
```
Comment (charles-cooper) [inline tests/unit/compiler/venom/test_multi_entry_block.py:137]: i'm not a huge fan of the hardcoded labels (i'd prefer to test properties on the cfg), but it's ok for now

File: vyper/venom/basicblock.py [hunk 31]
```diff
@@ -109,6 +125,12 @@ def __init__(self, value: int) -> None:
         assert isinstance(value, int), "value must be an int"
         self.value = value
 
+    def __hash__(self) -> int:
+        return self.value.__hash__()
+
+    def __eq__(self, v: object) -> bool:
+        return self.value == v
```
Comment (charles-cooper) [inline vyper/venom/basicblock.py:132]: :exploding_head: :exploding_head:

File: vyper/venom/basicblock.py [hunk 32]
```diff
@@ -131,14 +153,40 @@ class IRVariable(IRValue):
     mem_addr: Optional[int] = None
 
     def __init__(
-        self, value: str, mem_type: MemType = MemType.OPERAND_STACK, mem_addr: int = None
+        self,
+        value: str,
+        mem_type: MemType = MemType.OPERAND_STACK,
```
Comment (charles-cooper) [inline vyper/venom/basicblock.py:160]: add a note to remove the `mem_type` stuff (unless it's going to be useful later)

File: tests/unit/compiler/venom/test_dominator_tree.py (hunk 4 above)
Comment (charles-cooper) [reply tests/unit/compiler/venom/test_dominator_tree.py:73]:
```suggestion
    assert MakeSSA.run_pass(ctx, bb1) > 0
```

File: tests/functional/examples/factory/test_factory.py [hunk 33]
```diff
@@ -31,12 +31,14 @@ def create_exchange(token, factory):
...
-def factory(get_contract, optimize):
+def factory(get_contract, optimize, experimental_codegen):
     with open("examples/factory/Exchange.vy") as f:
         code = f.read()
 
     exchange_interface = vyper.compile_code(
-        code, output_formats=["bytecode_runtime"], settings=Settings(optimize=optimize)
+        code,
+        output_formats=["bytecode_runtime"],
+        settings=Settings(optimize=optimize, experimental_codegen=experimental_codegen),
```
[the remark this answers is not in the document]
Comment (charles-cooper) [reply tests/functional/examples/factory/test_factory.py:41]: can resolve later

**Review by charles-cooper — APPROVED** [review]
